nxcomp/Log.h

```diff
@@ -0,0 +1,502 @@
+		stream_             = &std::cerr;
+		level_              = NXDEBUG;
+		synchronized_       = true;
+		thread_buffer_size_ = 1024;
+		log_level_          = true;
+		log_time_           = true;
+		log_unix_time_      = false;
+		log_location_       = true;
+		log_thread_id_      = true;
+
+		if ( pthread_key_create(&tls_key_, free_thread_data) != 0 )
+		{
+			std::cerr << "pthread_key_create failed" << std::endl;
+			abort();
+		}
+
+	}
+
+	~NXLog()
+	{
+		per_thread_data *pdt = get_data();
+
+		// Flush any remaining output and delete TLS data
+		free_thread_data(pdt);
+
+		pthread_key_delete(tls_key_);
+	}
+
+	/** Minimum severity level to output */
+	NXLogLevel level() const
+	{
+		return level_;
+	}
+
+	void level(NXLogLevel level)
+	{
+		level_ = level;
+	}
+
+
+	/** Current severity level */
+	NXLogLevel current_level() const
+	{
+		return get_data()->current_level;
+	}
+
+	void current_level(NXLogLevel level)
+	{
+		get_data()->current_level = level;
+	}
+
+	/** Source file from which messages are currently originating */
+	std::string current_file() const
+	{
+		return *get_data()->current_file;
+	}
+
+	void current_file(std::string val)
+	{
+		*get_data()->current_file = val;
+	}
+
+	std::ostream* stream() const
+	{
+		return stream_;
+	}
+
+	void stream(std::ostream *stream)
+	{
+		flush();
+		stream_ = stream;
+	}
+
+	bool synchronized() const {
+		return synchronized_;
+	}
+
+	void synchronized(bool val) {
+		synchronized_ = val;
+	}
+
+	bool log_level() const
+	{
+		return log_level_;
+	}
+
+	void log_level(bool val)
+	{
+		log_level_ = val;
+	}
+
+	bool log_time() const 
+	{
+		return log_time_;
+	}
+
+	void log_time(bool val)
+	{
+		log_time_ = val;
+	}
+
+	bool log_unix_time() const
+	{
+		return log_unix_time_;
+	}
+
+	void log_unix_time(bool val)
+	{
+		log_unix_time_ = val;
+	}
+
+	bool log_location() const
+	{
+		return log_location_;
+	}
+
+	void log_location(bool val)
+	{
+		log_location_ = val;
+	}
+
+	bool log_thread_id() const
+	{
+		return log_thread_id_;
+	}
+
+	void log_thread_id(bool val)
+	{
+		log_thread_id_ = val;
+	}
+
+
+	void flush()
+	{
+		per_thread_data *pdt = get_data();
+		flush(pdt);
+	}
+
+	std::string thread_name() const
+	{
+		return *get_data()->thread_name;
+	}
+
+	void thread_name(std::string str)
+	{
+		*get_data()->thread_name = str;
+	}
+
+	void thread_name(const char *str)
+	{
+		*get_data()->thread_name = str;
+	}
+
+	/**
+	 * True if a message sent to the NXLog object will be sent to the output
+	 *
+	 * This considers two things:
+	 *
+	 * If there's a per-file log level, then it is used
+	 * Otherwise the global log level is used.
+	 *
+	 * If the log level permits the current message to be output, then the
+	 * return value is true.
+	 */ 
+	bool will_log() const;
+	
+
+	/**
+	 * This catches std::flush
+	 */
+	NXLog&  operator<<(std::ostream& (*F)(std::ostream&)) 
+	{
+		if ( will_log() )
+		{
+			if ( synchronized() )
+			{
+				per_thread_data *pdt = get_data();
+				(*pdt->buffer) << F;
```

Comment:
It's intended to also catch things like std::endl, which is a "\n" followed by a flush, and things like std::setprecision that change the state of the stream.

I think the mandatory flush() at the end isn't completely ideal, but it shouldn't do any harm. Stream manipulators are very rare in the code, and an extra flush is quite harmless, so it seemed to me like an OK compromise for the sake of simplicity.